src/adapter/rust_type_name.rs

```diff
@@ -0,0 +1,464 @@
+use std::fmt::{Display, Formatter, Result};
+
+/// Serializes the `rustdoc_types::Type` type as a String containing the name
+/// of the type, as close as possible to the original code declaration.
+pub(crate) fn rust_type_name(ty: &rustdoc_types::Type) -> String {
+    Type(ty).to_string()
+}
+
+/// Creates a struct named `$t` that wraps a `rustdoc_types::$t` reference,
+/// and implements `Display` on it by calling the given `$formatter` function.
+macro_rules! display_wrapper {
+    ($vis:vis $t:ident, $formatter:ident) => {
+        #[doc = concat!(
+            "Wraps a &rustdoc_types::",
+            stringify!($t),
+            " to implement [`Display`]."
+        )]
+        #[derive(Debug, Clone, Copy)]
+        $vis struct $t<'a>($vis &'a rustdoc_types::$t);
+
+        impl Display for $t<'_> {
+            fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+                $formatter(self, f)
+            }
+        }
+    };
+}
+
+/// Formats a sequence of `items` using the given formatter and format function `f`,
+/// putting the `delim` between each item (but with no leading or trailing delimiter).
+fn intersperse<T, I: IntoIterator<Item = T>, F: FnMut(T, &mut Formatter<'_>) -> Result>(
+    fmt: &mut Formatter<'_>,
+    delim: &str,
+    items: I,
+    mut f: F,
+) -> Result {
+    let mut first = true;
+    for item in items {
+        if first {
+            first = false;
+        } else {
+            write!(fmt, "{delim}")?;
+        }
+
+        f(item, fmt)?;
+    }
+
+    Ok(())
+}
+
+/// Calls [`intersperse`] with the given delimiter and items, using the `Display`
+/// impl of the value returned by the given function.
+fn intersperse_with<T, I: IntoIterator<Item = T>, U: Display, F: Fn(T) -> U>(
+    fmt: &mut Formatter<'_>,
+    delim: &str,
+    items: I,
+    f: F,
+) -> Result {
+    intersperse(fmt, delim, items, |x, fmt| f(x).fmt(fmt))
+}
+
+fn fmt_generic_param_def(this: &GenericParamDef, f: &mut Formatter<'_>) -> Result {
+    match &this.0.kind {
+        rustdoc_types::GenericParamDefKind::Lifetime { outlives } => {
+            write!(f, "{}", this.0.name)?;
+            if !outlives.is_empty() {
+                write!(f, ": ")?;
+                intersperse(f, " + ", outlives, String::fmt)?;
+            }
+        }
+        rustdoc_types::GenericParamDefKind::Type {
+            bounds,
+            default,
+            is_synthetic,
+        } => {
+            if *is_synthetic {
+                todo!("is_synthetic generic");
```

Comment:
If you'd like a test case for this, write a function that uses `impl Trait` in argument position.

In rustdoc JSON, you'll get a synthetic generic named `impl Trait` that corresponds to the unnamed generic type of that argument.

Reply:
I added a test case, but it's not triggering the `todo!`, so `is_synthetic` is not true here? I'm not sure.

Reply:
I think that nothing attempts to write out the generic parameters of the function that holds `impl Trait`. The test case prints out the types of the function parameters (and finds `impl Trait`), but the synthetic generic param shows up in the function's own type itself instead of the thing being printed and tested.

Try adding an additional function parameter with type like `impl FnOnce(&impl Clone) -> impl Clone`. When trying to print that parameter's type, that will run into (I believe) two synthetic type parameters, one for `&impl Clone` and one for `-> impl Clone`.

Desugared, that type is more-or-less `<T: Clone> impl FnOnce(&T) -> Self::R where Self::R: Clone`.

Reply:
```
error[E0562]: `impl Trait` is not allowed in the parameters of `Fn` trait bounds
  --> src/lib.rs:53:17
   |
53 |     f: impl Fn(&impl Clone) -> impl Clone,
   |                 ^^^^^^^^^^
   |
   = note: `impl Trait` is only allowed in arguments and return types of functions and methods

error[E0562]: `impl Trait` is not allowed in the return type of `Fn` trait bounds
  --> src/lib.rs:53:32
   |
53 |     f: impl Fn(&impl Clone) -> impl Clone,
   |                                ^^^^^^^^^^
   |
   = note: `impl Trait` is only allowed in arguments and return types of functions and methods
```

It looks like that's not allowed.  I'm going to search the JSON for when `is_synthetic` is true.

Reply:
I figured it out.  This is only for `GenericParamDef`, which we only use when we have a `for<...>` bound.  This currently only supports lifetimes, so this is not expressible (a `for<T> Fn(T)` bound is not currently valid).  This may come up when we are serializing generic definitions in functions, though.  We could mark the whole type branch as `unreachable!` for now.

Reply:
Yes, it's only in `GenericParamDef` which is why I was hoping to trigger it with `impl Fn(impl Trait)`, but I guess that won't work =/

I recently added [schema for generic parameters](https://github.com/obi1kenobi/trustfall-rustdoc-adapter/pull/514), among which some [test cases that do have synthetic generics](
https://github.com/obi1kenobi/trustfall-rustdoc-adapter/blob/941f2e6a7cb6123f97da851ce710a4c1da17b41f/test_crates/generic_parameters/src/lib.rs#L24-L27).

A related problem we have right now is that we can't write the Rust representation of `ImplementedTrait`, which shows up in trait bounds and in `impl SomeTrait<'a, T, 5> for MyType`.

Marking this unreachable or todo for now is fine then. If you're open to tackling the `ImplementedTrait` type printing problem next, that would be lovely and super useful because it'll improve the cargo-semver-checks lint diagnostics right away: instead of saying "type no longer implements `SomeTrait`", we can say it "no longer implements `for<'a> SomeTrait<'a, T, 5> where T: Clone + 'a`" — because it might still implement some other, incompatible variation of `SomeTrait`!

Reply:
We'll have the equivalent problem with generic parameters: a future lint we'll have is "this impl used to only require `T: Clone` but now it's also `T: Sync`"
```rust
// before
impl<T: Clone> SomeType<T> {
    pub fn example(value: T) {}
}

// after
impl<T: Clone + Sync> SomeType<T> {
    pub fn example(value: T) {}
}
```
You can't call `SomeType::example` with `T: !Sync` anymore (e.g. with `Rc<i64>`), so this is breaking. We'll need to print the affected generic parameter and its bounds (lifetimes, `use`, and `ImplementedTrait`)